### PR TITLE
Added exception handling when no elastic-search node is alive

### LINKED ISSUE #252: Catch ElasticSearch exceptions instead of throwing them in production
| Q | A
| --- | ---
| Bug? | no
| New Feature? | yes
| SuluArticleBundle Version | 0.7.0
| Sulu Version | 1.6.4

#### Actual Behavior

When there is an issue with the elasticsearch service which results in an exception being thrown, the page error 500s, taking down the whole site.

#### Expected Behavior

When there is an issue with the elasticsearch service in production environments, the exception should be logged and the search query should return null so that the rest of the site is still in operation.

#### Steps to Reproduce

Misconfigure elasticsearch parameters in production or turn the service off.

#### Possible Solutions

I had thought that a possible idea would be to dispatch an event when elastic search exceptions occur. This event could be handled by a listener that in dev/stage environments simply throws the exception, but in production has some logging and returns null. The rationale behind a listener/event implementation would be that it could easily be overwritten with custom error handling if needed. It could also be utilised in the main sulu bundle where elasticsearch maybe used for search results.

### PR DESCRIPTION
**Pull request by wachterjohannes**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | no
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | fixes #252 
| Related issues/PRs | none
| License | MIT

#### What's in this PR?

This PR introduces exception handling if elastic-search is not available.

#### Why?

The whole website crashes when no elastic-search is alive.

#### To Do

- [x] Add log-entry for missing elastic-search (warning)
- [x] Add documentation page

### CODE REVIEW

File: Content/ArticleDataProvider.php [hunk 1]
```diff
@@ -86,7 +95,8 @@ public function __construct(
         ReferenceStoreInterface $referenceStore,
         ArticleResourceItemFactory $articleResourceItemFactory,
         $articleDocumentClass,
-        $defaultLimit
+        $defaultLimit,
+        LoggerInterface $logger = null
```
Comment (alexander-schranz) [inline Content/ArticleDataProvider.php]: I personally would go the PSR way of injecting the logger over the `setLogger` using the LoggerAwareTrait:

https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-3-logger-interface.md

File: Content/ArticleDataProvider.php [hunk 2]
```diff
@@ -95,6 +105,7 @@ public function __construct(
         $this->articleResourceItemFactory = $articleResourceItemFactory;
         $this->articleDocumentClass = $articleDocumentClass;
         $this->defaultLimit = $defaultLimit;
+        $this->logger = $logger;
```
Comment (alexander-schranz) [inline Content/ArticleDataProvider.php]: when use the PSR way set the logger to `new NullLogger()` in the construct so it throws no error when the logger isn't injected.

File: Document/Repository/ArticleViewDocumentRepository.php [hunk 3]
```diff
@@ -57,19 +59,28 @@ class ArticleViewDocumentRepository
...
      * @param string $articleDocumentClass
      * @param array $searchFields
+     * @param LoggerInterface|null $logger
      */
     public function __construct(
         Manager $searchManager,
         $articleDocumentClass,
-        array $searchFields
+        array $searchFields,
+        LoggerInterface $logger = null
```
Comment (alexander-schranz) [inline Document/Repository/ArticleViewDocumentRepository.php]: same: https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-3-logger-interface.md

File: Markup/ArticleLinkProvider.php [hunk 4]
```diff
@@ -48,21 +50,29 @@ class ArticleLinkProvider implements LinkProviderInterface
...
      * @param string $articleViewClass
+     * @param LoggerInterface|null $logger
      */
     public function __construct(
         Manager $liveManager,
         Manager $defaultManager,
         array $types,
-        $articleViewClass
+        $articleViewClass,
+        LoggerInterface $logger = null
```
Comment (alexander-schranz) [inline Markup/ArticleLinkProvider.php]: same https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-3-logger-interface.md

File: Sitemap/ArticleSitemapProvider.php [hunk 5]
```diff
@@ -37,13 +40,23 @@ class ArticleSitemapProvider implements SitemapProviderInterface
...
-    public function __construct(Manager $manager, DocumentFactoryInterface $documentFactory)
-    {
+    public function __construct(
+        Manager $manager,
+        DocumentFactoryInterface $documentFactory,
+        LoggerInterface $logger = null
+    ) {
         $this->manager = $manager;
         $this->documentFactory = $documentFactory;
+        $this->logger = $logger;
```
Comment (alexander-schranz) [inline Sitemap/ArticleSitemapProvider.php]: same https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-3-logger-interface.md

File: Teaser/ArticleTeaserProvider.php [hunk 6]
```diff
@@ -43,15 +45,26 @@ class ArticleTeaserProvider implements TeaserProviderInterface
...
      * @param $articleDocumentClass
+     * @param LoggerInterface|null $logger
      */
-    public function __construct(Manager $searchManager, TranslatorInterface $translator, $articleDocumentClass)
-    {
+    public function __construct(
+        Manager $searchManager,
+        TranslatorInterface $translator,
+        $articleDocumentClass,
+        LoggerInterface $logger = null
```
Comment (alexander-schranz) [inline Teaser/ArticleTeaserProvider.php]: same https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-3-logger-interface.md

File: Sitemap/ArticleSitemapProvider.php [hunk 7]
```diff
@@ -117,7 +130,15 @@ private function getBulk(Repository $repository, $from, $size)
             ->setFrom($from)
             ->setSize($size);
 
-        return $repository->findDocuments($search);
+        try {
+            return $repository->findDocuments($search);
+        } catch (NoNodesAvailableException $exception) {
+            if ($this->logger) {
```
Comment (alexander-schranz) [inline Sitemap/ArticleSitemapProvider.php]: not needed when setting the logger to `new NullLogger()` in the construct.

File: Content/ArticleDataProvider.php [hunk 8]
```diff
@@ -249,7 +260,15 @@ private function getSearchResult(array $filters, $limit, $page, $pageSize, $loca
             $this->appendSortBy($filters['sortBy'], $sortMethod, $search);
         }
 
-        return $repository->findDocuments($search);
+        try {
+            return $repository->findDocuments($search);
+        } catch (NoNodesAvailableException $exception) {
+            if ($this->logger) {
```
Comment (alexander-schranz) [inline Content/ArticleDataProvider.php]: not needed when setting the logger to `new NullLogger()` in the construct

File: Content/ArticleSelectionContentType.php [hunk 9]
```diff
@@ -83,9 +93,19 @@ public function getContentData(PropertyInterface $property)
         $search = $repository->createSearch();
         $search->addQuery(new IdsQuery($this->getViewDocumentIds($value, $locale)));
 
+        try {
+            $documents = $repository->findDocuments($search);
+        } catch (NoNodesAvailableException $exception) {
+            if ($this->logger) {
```
Comment (alexander-schranz) [inline Content/ArticleSelectionContentType.php]: not needed when logger is `NullLogger` when no logger is given.

File: Document/Repository/ArticleViewDocumentRepository.php [hunk 10]
```diff
@@ -92,7 +103,15 @@ public function findRecent($excludeUuid = null, $limit = self::DEFAULT_LIMIT, ar
 
         $search->addSort(new FieldSort('authored', FieldSort::DESC));
 
-        return $this->repository->findDocuments($search);
+        try {
+            return $this->repository->findDocuments($search);
+        } catch (NoNodesAvailableException $exception) {
+            if ($this->logger) {
```
Comment (alexander-schranz) [inline Document/Repository/ArticleViewDocumentRepository.php]: same use `NullLogger`

File: Document/Repository/ArticleViewDocumentRepository.php [hunk 11]
```diff
@@ -121,7 +140,15 @@ public function findSimilar($uuid, $limit = self::DEFAULT_LIMIT, array $types =
             )
         );
 
-        return $this->repository->findDocuments($search);
+        try {
+            return $this->repository->findDocuments($search);
+        } catch (NoNodesAvailableException $exception) {
+            if ($this->logger) {
```
Comment (alexander-schranz) [inline Document/Repository/ArticleViewDocumentRepository.php]: use `NullLogger`

File: Markup/ArticleLinkProvider.php [hunk 12]
```diff
@@ -101,7 +111,15 @@ public function preload(array $hrefs, $locale, $published = true)
             $repository = $this->defaultManager->getRepository($this->articleViewClass);
         }
 
-        $documents = $repository->findDocuments($search);
+        try {
+            $documents = $repository->findDocuments($search);
+        } catch (NoNodesAvailableException $exception) {
+            if ($this->logger) {
```
Comment (alexander-schranz) [inline Markup/ArticleLinkProvider.php]: use `NullLogger`

File: Teaser/ArticleTeaserProvider.php [hunk 13]
```diff
@@ -169,19 +179,29 @@ public function find(array $ids, $locale)
         $search = $repository->createSearch();
         $search->addQuery(new IdsQuery($articleIds));
 
+        try {
+            $documents = $repository->findDocuments($search);
+        } catch (NoNodesAvailableException $exception) {
+            if ($this->logger) {
```
Comment (alexander-schranz) [inline Teaser/ArticleTeaserProvider.php]: use `NullLogger`

File: Content/ArticleDataProvider.php [hunk 14]
```diff
@@ -249,7 +257,13 @@ private function getSearchResult(array $filters, $limit, $page, $pageSize, $loca
             $this->appendSortBy($filters['sortBy'], $sortMethod, $search);
         }
 
-        return $repository->findDocuments($search);
+        try {
+            return $repository->findDocuments($search);
+        } catch (NoNodesAvailableException $exception) {
+            $this->logger->error($exception->getMessage());
```
Comment (alexander-schranz) [inline Content/ArticleDataProvider.php]: use `critical` log

**Comment by alexander-schranz** [comment]
maybe we should add a `view` parameter in the content types that he can show a error message in the frontend when the exception is catched.

**Comment by alexander-schranz** [comment]
ping @wachterjohannes

**Comment by wachterjohannes** [comment]
ping @sulu/core-team

**Comment by wachterjohannes** [comment]
we have decided to not handle this exceptions. the reason is that there is no "correct" way for this. we see elasticsearch as a hard dependency for the application like mysql which also will bring the application to crash if it is not there.

**Pull request closed without merging**